File: mem_file.c
#include "types.h"

#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <errno.h>

#include <sys/mman.h>
#include <fcntl.h>
#include <unistd.h>
#include <string.h>

#include "mem_file.h"

//index in int(4 byte size)
#define ITEM_SIZE   0
#define ITEM_ITEMS  1
#define ITEM_INDEX  2
#define ITEM_TYPE   3

//index in byte(1 byte size)
#define SEMAPHORE_POS   14

char *mmap_file(char * filename, int size) // return char* to buffer
{
    int fd = open(filename, O_RDWR | O_CREAT, S_IRUSR | S_IWUSR );//shm_open(STORAGE_ID, O_RDWR, S_IRUSR | S_IWUSR);
    if (fd < 0)
    {
        printf("open issue for %s:\n", filename);
        perror("open");
        return 0;
    }
    // allocate size in file
    if(ftruncate(fd, size+10) != 0) 
    {
        printf("ERROR: ftruncate issue\n");
        return 0;
    }
    char *buffer = mmap_fd_write(fd, size);
    close(fd);
    return buffer;
}

char *mmap_fd_write(int fd, int size) // return char* to buffer
{ 
    errno = 0;
    // mmap file
    char *buffer = mmap(NULL, size, PROT_READ | PROT_WRITE, MAP_SHARED | MAP_POPULATE, fd, 0);
    if (buffer == MAP_FAILED)
    {
        perror("mmap");
        return 0;
    }
    int i;
    volatile char val;
    for(i = 0; i < size; i++)
    {
        val = buffer[i];
    }
    perror("mmap");
    return buffer;
}

char *mmap_fd_read(int fd, int size) // return char* to buffer
{ 
    errno = 0;
    // mmap file
    char *buffer = mmap(NULL, size, PROT_READ, MAP_SHARED | MAP_POPULATE, fd, 0);
    if (buffer == MAP_FAILED)
    {
        perror("mmap");
        return 0;
    }
    int i;
    volatile char val;
    for(i = 0; i < size; i++)
    {
        val = buffer[i];
    }
    perror("mmap");
    return buffer;
}

int munmap_file(char * buffer, int size) // return char* to buffer
{
    return munmap(buffer, size);
}

int calculate_buffer_size(int item_size, int max_items)
{
    return 16 + (item_size * max_items);
}

int calculate_buffer_size_from_file(int fd) //calulate buffer size before the file is mmapped
{
    int item_size = 0;
    if(read(fd,(char *)&item_size,4) != 4)
    {
        printf("ERROR: could not read item size");
        return -1;
    }

    int max_items = 0;
    if(read(fd,(char *)&max_items,4) != 4)
    {
        printf("ERROR: could not read max items");
        return -1;
    }
    //int index = read(fd,4);
    //int semaphore = read(fd,2);
    //int type = read(fd,2);
    return 16 + item_size*max_items;
}

inline int type_to_item_size(int type)
{
    switch(type)
    {
        case 0:
            return 0;
        case BOOL:
            return 1;
        case INT8:
            return 1;
        case INT32:
            return 4;
        case INT64:
            return 8;
        case FLOAT32:
            return 4;
        case FLOAT64:
            return 8;
        case QUALITY:
            return 4;
        case SMV92:
            return 68;
        default:
            return 0;
    }
}

inline int read_size(char * buffer)
{
    int *tmp = (int *)buffer;
    return tmp[0];
}

inline int read_items(char * buffer)
{
    int *tmp = (int *)buffer;
    return tmp[1];
}

inline int read_index(char * buffer)
{
    int *tmp = (int *)buffer;
    return tmp[2];
}

inline int read_type(char * buffer)
{
    int *tmp = (int *)buffer;
    return tmp[3] & 0x0000FFFF;
}


inline int lock_buffer(volatile char * buffer, char spin) //TODO: does volatile here really prevent optimisation of the busy-wait loop???
{
    if (spin == 0)//test or wait
    {
        if(buffer[SEMAPHORE_POS])
        {
            return -1;
        }
    }
    else
    {
        while(buffer[SEMAPHORE_POS]); //busy wait
    }
    buffer[SEMAPHORE_POS] = 1; //write semaphore to 1
    return 0;   
}

inline int free_buffer_lock(char * buffer)
{
    if(!buffer[SEMAPHORE_POS])//if already 0
    {
        return -1;
    }
    buffer[SEMAPHORE_POS] = 0; //write semaphore to 0
    return 0; 
}

inline char read_input_bool(char * buffer, int index)
{
    lock_buffer(buffer,1);
    register char result = buffer[index + 16];
    free_buffer_lock(buffer);
    return result;
}

inline char read_input_int8(char * buffer, int index)
{
    lock_buffer(buffer,1);
    register char result =  buffer[index + 16];
    free_buffer_lock(buffer);
    return result;
}

inline int read_input_int32(char * buffer, int index)
{
    int *tmp = (int *)buffer;
    lock_buffer(buffer,1);
    register int result = tmp[index + 4];
    free_buffer_lock(buffer);
    return result;
}

inline int read_input_smv(char * buffer, int var_index, int index)
{
    int *tmp = (int *)buffer;
    lock_buffer(buffer,1);
    register int result = tmp[4+ (index * 17) + var_index];
    free_buffer_lock(buffer);
    return result;
}

inline char read_current_input_bool(char * buffer)
{
    lock_buffer(buffer,1);
    register char result = buffer[read_index(buffer) + 16];
    free_buffer_lock(buffer);
    return result;
}

inline char read_current_input_int8(char * buffer)
{
    lock_buffer(buffer,1);
    register char result =  buffer[read_index(buffer) + 16];
    free_buffer_lock(buffer);
    return result;
}

inline int read_current_input_int32(char * buffer)
{
    int *tmp = (int *)buffer;
    lock_buffer(buffer,1);
    register int result = tmp[tmp[2] + 4];//tmp[2] == index
    free_buffer_lock(buffer);
    return result;
}

inline int write_type(char *buffer, short type)
{
    short *tmp = (short *)buffer;
    lock_buffer(buffer,1);
    tmp[6] = type;
    free_buffer_lock(buffer);
}

inline int write_size(char *buffer, int size)
{
    int *tmp = (int *)buffer;
    lock_buffer(buffer,1);
    tmp[0] = size;
    free_buffer_lock(buffer);
}

inline int write_items(char *buffer, int items)
{
    int *tmp = (int *)buffer;
    lock_buffer(buffer,1);
    tmp[1] = items;
    free_buffer_lock(buffer);
}

inline int write_index(char *buffer, int index)
{
    int *tmp = (int *)buffer;
    lock_buffer(buffer,1);
    tmp[2] = index;
    free_buffer_lock(buffer);
}

inline int write_output_int(char * buffer, int value)
{
    int *tmp = (int *)buffer;
    if(read_type(buffer) != INT32 || read_size(buffer) != 4) return -1;
    int items = read_items(buffer);

    lock_buffer(buffer,1);

    int index = (read_index(buffer) + 1) % items;
    tmp[ 4 + index] = value;
    tmp[2] = index; //increment index

    free_buffer_lock(buffer);
}

int write_output_bool(char * buffer, char value)
{
    int *tmp = (int *)buffer;
    if(read_type(buffer) != BOOL || read_size(buffer) != 1) return -1;
    int items = read_items(buffer);

    lock_buffer(buffer,1);

    int index = (read_index(buffer) + 1) % items;
    buffer[ 16 + index] = value;
    tmp[2] = index; //increment index

    free_buffer_lock(buffer);
}

int write_output_int8(char * buffer, char value)
{
    int *tmp = (int *)buffer;
    if(read_type(buffer) != INT8 || read_size(buffer) != 1) return -1;
    int items = read_items(buffer);

    lock_buffer(buffer,1);

    int index = (read_index(buffer) + 1) % items;
    buffer[ 16 + index] = value;
    tmp[2] = index; //increment index

    free_buffer_lock(buffer);
}

int write_output_quality(char * buffer, int value)
{
    int *tmp = (int *)buffer;
    if(read_type(buffer) != QUALITY || read_size(buffer) != 4) return -1;
    int items = read_items(buffer);

    lock_buffer(buffer,1);

    int index = (read_index(buffer) + 1) % items;
    tmp[ 4 + index] = value;
    tmp[2] = index; //increment index

    free_buffer_lock(buffer);
}

int write_output_time(char * buffer, int value)
{
    int *tmp = (int *)buffer;
    if(read_type(buffer) != GENTIME || read_size(buffer) != 4) return -1;
    int items = read_items(buffer);

    lock_buffer(buffer,1);

    int index = (read_index(buffer) + 1) % items;
    tmp[ 4 + index] = value;
    tmp[2] = index; //increment index

    free_buffer_lock(buffer);
}

int write_output_float(char * buffer, float value)
{
    float *tmp = (float *)buffer;
    if(read_type(buffer) != FLOAT32 || read_size(buffer) != 4) return -1;
    int items = read_items(buffer);

    lock_buffer(buffer,1);

    int index = (read_index(buffer) + 1) % items;
    tmp[ 4 + index] = value;
    tmp[2] = index; //increment index

    free_buffer_lock(buffer);
}

File: other/python_example/mem_file.py
import struct
import sys

#https://docs.python.org/3/library/struct.html

# constants
SIZE_POS=0
ITEMS_POS=4
INDEX_POS=8
TYPE_POS=12
SEMAPHORE_POS=14
DATA_POS=16

# types
BOOL  =      1  # item size = 1
INT8  =      2  # item size = 1
INT32 =      3  # item size = 4
INT64 =      4  # item size = 8
FLOAT32 =    5  # item size = 4
FLOAT64 =    6  # item size = 8
QUALITY =    7  # item size = 4
MMSSTRING =  8  # item size = max strlen
BITSTRING =  9  # item size is then in bits instead of bytes
VISSTRING =  10 # item size = max strlen
OCTSTRING =  11 # item size = max strlen
GENTIME =    12 # item size = 4?
BINTIME =    13 # item size = 4?
UTCTIME =    14 # item size = 4?
GOOSE =      20 # items size 1024...
SMV92 =      68 # item size = 64

# item sizes
def type_to_item_size(type):
    switch = {
        "BOOL":1,
        "INT8":1,
        "INT32":4,
        "INT64":8,
        "FLOAT32":4,
        "FLOAT64":8,
        "QUALITY":4,
        "SMV92":68,
        "GOOSE":1024,
    }
    return switch.get(type,"unknown")


def read_size(file):
    file.seek(SIZE_POS, 0)
    return int(struct.unpack('i', file.read(4))[0])


def read_items(file):
    file.seek(ITEMS_POS, 0)
    return int(struct.unpack('i', file.read(4))[0])


def read_index(file):
    file.seek(INDEX_POS, 0)
    return int(struct.unpack('i', file.read(4))[0])


def read_type(file):
    file.seek(TYPE_POS, 0)
    return int(struct.unpack('h', file.read(2))[0])


def lock_buffer(file, spin):
    old = file.tell()
    file.seek(SEMAPHORE_POS, 0)
    if spin == False:
        if file.read(1) == 1:
            file.seek(old,0)
            return -1
    else:
        while file.read(1) == 1: #busy wait loop
            file.seek(SEMAPHORE_POS, 0)
            
    file.seek(SEMAPHORE_POS, 0)
    file.write( bytes(1) )
    file.seek(old,0)
    return 0


def free_buffer_lock(file):
    old = file.tell()
    file.seek(SEMAPHORE_POS, 0)
    if file.read(1) == 0:
        file.seek(old,0)
        return -1

    file.seek(SEMAPHORE_POS, 0)
    file.write( bytes(0) )
    file.seek(old,0)
    return 0


def read_input_bool(file, index):
    file.seek(DATA_POS + index, 0)
    lock_buffer(file, True)
    result = bool(file.read(1))
    free_buffer_lock(file)
    return result

def read_input_int8(file, index):
    file.seek(DATA_POS + index, 0)
    lock_buffer(file, True)
    result = int(file.read(1))
    free_buffer_lock(file)
    return result

def read_input_int32(file, index):
    file.seek(DATA_POS + (index * 4), 0)
    lock_buffer(file, True)
    result = int(struct.unpack('i', file.read(4)))
    free_buffer_lock(file)
    return result

def read_input_smv(file, var_index, index):
    file.seek(DATA_POS + (index * 68) + var_index, 0)
    lock_buffer(file, True)
    result = int(struct.unpack('i', file.read(4)))
    free_buffer_lock(file)
    return result


def write_type(file, type):
    file.seek(TYPE_POS, 0)
    lock_buffer(file, True)
    file.write( bytes(struct.pack("i", type)) )
    free_buffer_lock(file)


def write_size(file, size):
    file.seek(SIZE_POS, 0)
    lock_buffer(file, True)
    file.write( bytes(struct.pack("i", size)) )
    free_buffer_lock(file)


def write_items(file, items):
    file.seek(ITEMS_POS, 0)
    lock_buffer(file, True)
    file.write( bytes(struct.pack("i", items)) )
    free_buffer_lock(file)


def write_index(file, index):
    file.seek(INDEX_POS, 0)
    lock_buffer(file, True)
    file.write( bytes(struct.pack("i", index)) )
    free_buffer_lock(file)


def write_output_int(file, value):
    if read_type(file) != INT32 or read_size(file) != 4:
        return -1

    items = read_items(file);

    lock_buffer(file,1);

    index = (read_index(file) + 1) % items;

    file.seek(DATA_POS + (index * 4), 0)
    file.write( bytes(struct.pack("i", value)) )

    write_index(file,index)

    free_buffer_lock(file);


def write_output_bool(file, value):
    if read_type(file) != BOOL or read_size(file) != 1:
        return -1

    items = read_items(file);

    lock_buffer(file,1);

    index = (read_index(file) + 1) % items;

    file.seek(DATA_POS + (index), 0)
    file.write( bytes(struct.pack("?", value)) )

    write_index(file,index)

    free_buffer_lock(file);


def write_output_int8(file, value):
    if read_type(file) != INT8 or read_size(file) != 1:
        return -1

    items = read_items(file);

    lock_buffer(file,1);

    index = (read_index(file) + 1) % items;

    file.seek(DATA_POS + (index), 0)
    file.write( bytes(struct.pack("b", value)) )

    write_index(file,index)

    free_buffer_lock(file);



def write_output_quality(file, value):
    if read_type(file) != QUALITY or read_size(file) != 4:
        return -1

    items = read_items(file);

    lock_buffer(file,1);

    index = (read_index(file) + 1) % items;

    file.seek(DATA_POS + (index * 4), 0)
    file.write( bytes(struct.pack("i", value)) )

    write_index(file,index)

    free_buffer_lock(file);


def write_output_time(file, value):
    if read_type(file) != GENTIME or read_size(file) != 4:
        return -1

    items = read_items(file);

    lock_buffer(file,1);

    index = (read_index(file) + 1) % items;

    file.seek(DATA_POS + (index * 4), 0)
    file.write( bytes(struct.pack("i", value)) )

    write_index(file,index)

    free_buffer_lock(file);


def write_output_float(file, value):
    if read_type(file) != FLOAT32 or read_size(file) != 4:
        return -1

    items = read_items(file);

    lock_buffer(file,1);

    index = (read_index(file) + 1) % items;

    file.seek(DATA_POS + (index * 4), 0)
    file.write( bytes(struct.pack("f", value)) )

    write_index(file,index)

    free_buffer_lock(file);


def read_current_input_bool(file):
    index = read_index(file)
    return read_input_bool(file, index)


def read_current_input_int8(file):
    index = read_index(file)
    return read_input_int8(file, index)


def read_current_input_int32(file):
    index = read_index(file)
    return read_input_int32(file, index)


#testing
file = open("/dev/shm/smv92.mem", "r+b")
size = read_size(file)
items = read_items(file)
index = read_index(file)
type = read_type(file)
print("size=%i, items=%i, index=%i, type=%i" % (size,items, index, type))

result = lock_buffer(file,True)
print("lock result= %i" % result)
result = free_buffer_lock(file)
print("unlock result= %i" % result)

print("done")


